File: dynamic_tf2_publisher_py/dynamic_tf2_publisher_py/dynamic_tf2_publisher.py
import math
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TransformStamped
from tf2_ros import TransformBroadcaster

class DynamicTF2Publisher(Node):
    def __init__(self):
        super().__init__("dynamic_tf2_publisher")

        # ----------------------------------------------------
        # DÉCLARER LES PARAMÈTRES
        # ----------------------------------------------------
        self.declare_parameter("parent_frame", "map")
        self.declare_parameter("child_frame",  "base_link")
        self.declare_parameter("x",   0.0)
        self.declare_parameter("y",   0.0)
        self.declare_parameter("z",   0.0)
        self.declare_parameter("yaw", 0.0)

        # Lire les paramètres
        self.parent_frame_ = self.get_parameter("parent_frame").get_parameter_value().string_value
        self.child_frame_  = self.get_parameter("child_frame").get_parameter_value().string_value
        self.x_   = self.get_parameter("x").get_parameter_value().double_value
        self.y_   = self.get_parameter("y").get_parameter_value().double_value
        self.z_   = self.get_parameter("z").get_parameter_value().double_value
        self.yaw_ = self.get_parameter("yaw").get_parameter_value().double_value

        # ----------------------------------------------------
        # CRÉER LE BROADCASTER TF2
        # ----------------------------------------------------
        self.tf_broadcaster_ = TransformBroadcaster(self)

        # Angle courant pour l'animation
        self.angle_ = 0.0

        # Timer : publier toutes les 100ms
        self.timer_ = self.create_timer(0.1, self.timer_callback)

        self.get_logger().info(
            f"TF2 Publisher started : {self.parent_frame_} → {self.child_frame_}"
        )

    def timer_callback(self):
        # ----------------------------------------------------
        # CRÉER LE MESSAGE DE TRANSFORMATION
        # ----------------------------------------------------
        transform = TransformStamped()

        # En-tête
        transform.header.stamp    = self.get_clock().now().to_msg()
        transform.header.frame_id = self.parent_frame_
        transform.child_frame_id  = self.child_frame_

        # ----------------------------------------------------
        # TRANSLATION
        # ----------------------------------------------------
        transform.transform.translation.x = self.x_
        transform.transform.translation.y = self.y_
        transform.transform.translation.z = self.z_

        # ----------------------------------------------------
        # ROTATION (quaternion depuis yaw)
        # ----------------------------------------------------
        transform.transform.rotation.x = 0.0
        transform.transform.rotation.y = 0.0
        transform.transform.rotation.z = math.sin(self.angle_ / 2.0)
        transform.transform.rotation.w = math.cos(self.angle_ / 2.0)

        # Animation : faire tourner lentement
        self.angle_ += 0.01
        if self.angle_ > 2 * math.pi:
            self.angle_ = 0.0

        # Publier la transformation
        self.tf_broadcaster_.sendTransform(transform)

def main(args=None):
    rclpy.init(args=args)
    node = DynamicTF2Publisher()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: ros2_fundamentals/dynamic_tf2_publisher/src/dynamic_tf2_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
//       ↑ message pour représenter une transformation

#include "tf2_ros/transform_broadcaster.h"
//       ↑ bibliothèque pour publier des transformations TF2

#include <cmath>
//       ↑ pour les fonctions mathématiques (sin, cos...)

class DynamicTF2Publisher : public rclcpp::Node
{
public:
    DynamicTF2Publisher() : Node("dynamic_tf2_publisher"), angle_(0.0)
    {
        // ----------------------------------------------------
        // DÉCLARER LES PARAMÈTRES
        // ----------------------------------------------------
        this->declare_parameter("parent_frame", std::string("map"));
        this->declare_parameter("child_frame",  std::string("base_link"));
        this->declare_parameter("x",    0.0);
        this->declare_parameter("y",    0.0);
        this->declare_parameter("z",    0.0);
        this->declare_parameter("yaw",  0.0);

        // Lire les paramètres
        parent_frame_ = this->get_parameter("parent_frame").as_string();
        child_frame_  = this->get_parameter("child_frame").as_string();
        x_   = this->get_parameter("x").as_double();
        y_   = this->get_parameter("y").as_double();
        z_   = this->get_parameter("z").as_double();
        yaw_ = this->get_parameter("yaw").as_double();

        // ----------------------------------------------------
        // CRÉER LE BROADCASTER TF2
        // ----------------------------------------------------
        tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
        //                  ↑ publie les transformations sur /tf

        // Timer : publier la transformation toutes les 100ms
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100),
            std::bind(&DynamicTF2Publisher::timerCallback, this)
        );

        RCLCPP_INFO(this->get_logger(),
            "TF2 Publisher started : %s → %s",
            parent_frame_.c_str(), child_frame_.c_str());
    }

private:
    void timerCallback()
    {
        // ----------------------------------------------------
        // CRÉER LE MESSAGE DE TRANSFORMATION
        // ----------------------------------------------------
        geometry_msgs::msg::TransformStamped transform;

        // En-tête : timestamp et frame parent
        transform.header.stamp    = this->get_clock()->now();
        //                           ↑ timestamp actuel
        transform.header.frame_id = parent_frame_;
        //                           ↑ frame de référence (parent)

        // Frame enfant
        transform.child_frame_id = child_frame_;

        // ----------------------------------------------------
        // TRANSLATION (position x, y, z)
        // ----------------------------------------------------
        transform.transform.translation.x = x_;
        transform.transform.translation.y = y_;
        transform.transform.translation.z = z_;

        // ----------------------------------------------------
        // ROTATION (quaternion)
        // TF2 utilise des quaternions pour les rotations
        // On convertit l'angle yaw en quaternion
        // ----------------------------------------------------
        // Pour une rotation autour de Z (yaw) :
        // w = cos(angle/2)
        // z = sin(angle/2)
        // x = y = 0
        transform.transform.rotation.x = 0.0;
        transform.transform.rotation.y = 0.0;
        transform.transform.rotation.z = std::sin(angle_ / 2.0);
        transform.transform.rotation.w = std::cos(angle_ / 2.0);

        // Faire tourner lentement pour visualiser
        angle_ += 0.01;
        if (angle_ > 2 * M_PI) angle_ = 0.0;
        //                 ↑ remettre à zéro après un tour complet

        // Publier la transformation
        tf_broadcaster_->sendTransform(transform);
    }

    std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
    rclcpp::TimerBase::SharedPtr timer_;

    std::string parent_frame_;
    std::string child_frame_;
    double x_, y_, z_, yaw_;
    double angle_;   // angle courant pour l'animation
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<DynamicTF2Publisher>());
    rclcpp::shutdown();
    return 0;
}